File: BadgeGenerator.cs
// BadgeGenerator.cs
using System;
using System.IO;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;
using SixLabors.ImageSharp.Drawing.Processing;
using SixLabors.Fonts;

namespace BadgeGenerator
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            if (string.IsNullOrEmpty(opts.Name) || string.IsNullOrEmpty(opts.Photo))
            {
                Console.Error.WriteLine("Ошибка: --name и --photo обязательны");
                return;
            }
            var gen = new BadgeGenerator(opts);
            gen.Generate(opts.Output);
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--name": opts.Name = args[++i]; break;
                    case "--title": opts.Title = args[++i]; break;
                    case "--company": opts.Company = args[++i]; break;
                    case "--photo": opts.Photo = args[++i]; break;
                    case "--output": opts.Output = args[++i]; break;
                    case "--width": opts.Width = int.Parse(args[++i]); break;
                    case "--height": opts.Height = int.Parse(args[++i]); break;
                    case "--bg": opts.Bg = args[++i]; break;
                    case "--text-color": opts.TextColor = args[++i]; break;
                    case "--photo-size": opts.PhotoSize = int.Parse(args[++i]); break;
                    case "--photo-shape": opts.PhotoShape = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string Name { get; set; }
            public string Title { get; set; }
            public string Company { get; set; }
            public string Photo { get; set; }
            public string Output { get; set; } = "badge.png";
            public int Width { get; set; } = 600;
            public int Height { get; set; } = 400;
            public string Bg { get; set; } = "#FFFFFF";
            public string TextColor { get; set; } = "#000000";
            public int PhotoSize { get; set; } = 120;
            public string PhotoShape { get; set; } = "circle";
        }

        class BadgeGenerator
        {
            private Options opts;

            public BadgeGenerator(Options opts)
            {
                this.opts = opts;
            }

            public void Generate(string output)
            {
                using (Image<Rgba32> badge = new Image<Rgba32>(opts.Width, opts.Height))
                {
                    // Фон
                    var bgColor = Color.ParseHex(opts.Bg);
                    badge.Mutate(ctx => ctx.BackgroundColor(bgColor));

                    // Фото
                    using (Image<Rgba32> photo = Image.Load<Rgba32>(opts.Photo))
                    {
                        photo.Mutate(x => x.Resize(opts.PhotoSize, opts.PhotoSize));
                        if (opts.PhotoShape == "circle")
                        {
                            // Создаём круглую маску (упрощённо: применяем clipping)
                            photo.Mutate(x => x.Crop(new EllipsePolygon(opts.PhotoSize/2, opts.PhotoSize/2, opts.PhotoSize/2)));
                        }
                        int photoX = 30;
                        int photoY = (opts.Height - opts.PhotoSize) / 2;
                        badge.Mutate(ctx => ctx.DrawImage(photo, new Point(photoX, photoY), 1f));
                    }

                    // Текст
                    var textColor = Color.ParseHex(opts.TextColor);
                    var font = SystemFonts.CreateFont("Arial", 36, FontStyle.Bold);
                    int textX = 30 + opts.PhotoSize + 30;
                    int textY = (opts.Height - 90) / 2;
                    badge.Mutate(ctx => ctx.DrawText(opts.Name, font, textColor, new PointF(textX, textY)));
                    if (!string.IsNullOrEmpty(opts.Title))
                    {
                        var fontTitle = SystemFonts.CreateFont("Arial", 24, FontStyle.Regular);
                        badge.Mutate(ctx => ctx.DrawText(opts.Title, fontTitle, textColor, new PointF(textX, textY + 40)));
                    }
                    if (!string.IsNullOrEmpty(opts.Company))
                    {
                        var fontCompany = SystemFonts.CreateFont("Arial", 28, FontStyle.Regular);
                        badge.Mutate(ctx => ctx.DrawText(opts.Company, fontCompany, textColor, new PointF(textX, textY + 70)));
                    }

                    // Сохраняем
                    badge.Save(output);
                    Console.WriteLine($"Бейдж сохранён в {output}");
                }
            }
        }
    }
}

File: badge_generator.js
#!/usr/bin/env node
// badge_generator.js
const { program } = require('commander');
const fs = require('fs');
const sharp = require('sharp');
const chalk = require('chalk');

class BadgeGenerator {
    constructor(options) {
        this.name = options.name;
        this.title = options.title || '';
        this.company = options.company || '';
        this.photoPath = options.photo;
        this.width = options.width || 600;
        this.height = options.height || 400;
        this.bgColor = options.bg || '#FFFFFF';
        this.textColor = options.textColor || '#000000';
        this.photoSize = options.photoSize || 120;
        this.photoShape = options.photoShape || 'circle';
        this.font = options.font || null;
    }

    async generate(output) {
        // Создаём фон
        let badge = sharp({
            create: {
                width: this.width,
                height: this.height,
                channels: 3,
                background: this.bgColor
            }
        });

        // Обрабатываем фото
        let photo = sharp(this.photoPath);
        const photoMeta = await photo.metadata();
        let photoBuffer = await photo
            .resize(this.photoSize, this.photoSize, { fit: 'cover' })
            .toBuffer();

        // Если круг, создаём маску
        if (this.photoShape === 'circle') {
            const mask = await sharp({
                create: {
                    width: this.photoSize,
                    height: this.photoSize,
                    channels: 1,
                    background: { r: 0, g: 0, b: 0 }
                }
            })
            .composite([{
                input: await sharp({
                    create: {
                        width: this.photoSize,
                        height: this.photoSize,
                        channels: 1,
                        background: { r: 255, g: 255, b: 255 }
                    }
                })
                .composite([{
                    input: Buffer.from(`<svg><circle cx="${this.photoSize/2}" cy="${this.photoSize/2}" r="${this.photoSize/2}" fill="black"/></svg>`),
                    blend: 'dest-in'
                }])
                .png()
                .toBuffer()
            }])
            .png()
            .toBuffer();
            photoBuffer = await sharp(photoBuffer)
                .composite([{ input: mask, blend: 'dest-in' }])
                .png()
                .toBuffer();
        }

        // Накладываем фото на бейдж
        const photoX = 30;
        const photoY = (this.height - this.photoSize) / 2;
        badge = badge.composite([{
            input: photoBuffer,
            top: Math.round(photoY),
            left: photoX
        }]);

        // Создаём SVG для текста (для простоты используем SVG-оверлей)
        const textX = photoX + this.photoSize + 30;
        const textY = (this.height - 90) / 2;
        let svg = `<svg width="${this.width}" height="${this.height}">`;
        svg += `<text x="${textX}" y="${textY}" font-size="36" fill="${this.textColor}" font-family="Arial, sans-serif">${this.name}</text>`;
        if (this.title) {
            svg += `<text x="${textX}" y="${textY + 40}" font-size="24" fill="${this.textColor}" font-family="Arial, sans-serif">${this.title}</text>`;
        }
        if (this.company) {
            svg += `<text x="${textX}" y="${textY + 70}" font-size="28" fill="${this.textColor}" font-family="Arial, sans-serif">${this.company}</text>`;
        }
        svg += '</svg>';

        badge = badge.composite([{
            input: Buffer.from(svg),
            top: 0,
            left: 0
        }]);

        await badge.toFile(output);
        console.log(chalk.green(`Бейдж сохранён в ${output}`));
    }
}

program
    .option('--name <name>', 'Имя на бейдже')
    .option('--title <title>', 'Должность')
    .option('--company <company>', 'Компания')
    .requiredOption('--photo <path>', 'Путь к фото')
    .option('--output <file>', 'Выходной файл', 'badge.png')
    .option('--width <number>', 'Ширина', parseInt, 600)
    .option('--height <number>', 'Высота', parseInt, 400)
    .option('--bg <color>', 'Цвет фона', '#FFFFFF')
    .option('--text-color <color>', 'Цвет текста', '#000000')
    .option('--photo-size <number>', 'Размер фото', parseInt, 120)
    .option('--photo-shape <shape>', 'Форма фото (circle, square)', 'circle')
    .option('--font <path>', 'Путь к шрифту (не используется в SVG)')
    .option('--batch <file>', 'CSV для пакетной генерации')
    .option('--output-dir <dir>', 'Директория для пакетного вывода', '.')
    .parse(process.argv);

const opts = program.opts();

if (opts.batch) {
    const csv = require('csv-parser');
    const fs = require('fs');
    const path = require('path');
    const stream = fs.createReadStream(opts.batch).pipe(csv());
    stream.on('data', async (row) => {
        const name = row.name || '';
        const title = row.title || '';
        const company = row.company || '';
        const photo = row.photo || '';
        const out = row.output || `${name.replace(/\s/g, '_')}.png`;
        const outPath = path.join(opts.outputDir, out);
        const gen = new BadgeGenerator({
            name, title, company, photo,
            width: opts.width,
            height: opts.height,
            bg: opts.bg,
            textColor: opts.textColor,
            photoSize: opts.photoSize,
            photoShape: opts.photoShape,
            font: opts.font
        });
        try {
            await gen.generate(outPath);
        } catch (e) {
            console.error(chalk.red(`Ошибка при генерации ${outPath}: ${e.message}`));
        }
    });
} else {
    if (!opts.name) {
        console.error(chalk.red('Ошибка: --name обязателен'));
        process.exit(1);
    }
    const gen = new BadgeGenerator(opts);
    gen.generate(opts.output);
}
